sprint_7-solution-in-file-manager

devices11 opened the pull request:
1. Добавлен менеджер с сохранением всех задач и их состояний
2. Добавлены тесты

Commit:
1. Переименован метод получения истории
2. В тестах добавлены пустые строки для удобства чтения

Commit:
1. Добавлен менеджер с сохранением всех задач и их состояний
2. Добавлены тесты

Commit:
Почищены неиспользуемые импорты и переменные

Ivanov-s-tmn commented on test/service/FileBackedTaskManagerTest.java:
```diff
@@ -0,0 +1,191 @@
+
+    @BeforeEach
+    public void beforeEach() {
+        try {
+            file = File.createTempFile("storage", ".csv");
+            taskManager = FileBackedTaskManager.loadFromFile(Managers.getDefaultHistory(), file);
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
```
В тестах обычно не обрабатываются исключения, достаточно указать в сигнатуре метода throws

Ivanov-s-tmn commented on src/main/service/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,217 @@
+        return task.getId() + "," +
+                type + "," +
+                task.getTitle() + "," +
+                task.getStatus() + "," +
+                task.getDescription() + "," +
+                epicId + "\n";
+    }
+
+    //Преобразовать строку в объект задачи
+    private Task fromString(String value) {
+        String[] splitValue = value.split(",");
+        int id = Integer.parseInt(splitValue[0]);
+        TypeTask type = valueOf(splitValue[1]);
+        String name = splitValue[2];
+        StatusModel status = StatusModel.valueOf(splitValue[3]);
+        String description = splitValue[4];
+        int epicId = 0;
+        if (type.equals(SUBTASK)) {
+            epicId = Integer.parseInt(splitValue[5]);
+        }
+
+        if (type == TASK) {
+            Task task = new Task(name, description);
+            task.setId(id);
+            task.setStatus(status);
+            return task;
+        } else if (type == EPIC) {
+            Epic epic = new Epic(name, description);
+            epic.setId(id);
+            epic.setStatus(status);
+            return epic;
+        } else if (type == SUBTASK) {
+            Subtask subtask = new Subtask(name, description, epicId);
+            subtask.setId(id);
+            subtask.setStatus(status);
+            return subtask;
+        }
+
+        return null;
+    }
```
Методы преобразования из строки/в строку лучше вынести в отдельные утилитный класс, чтобы разгрузить файловый менеджер

Ivanov-s-tmn commented on src/main/service/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,217 @@
+            return subtask;
+        }
+
+        return null;
```
Явно возвращать из метода null не лучшая практика, у нас есть 3 типа задач, которые могут находиться в файле и других типов пока не предусмотрено

Ivanov-s-tmn commented on src/main/service/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,217 @@
+    private String toString(Task task) {
+        TypeTask type = null;
+        String epicId = "";
+        if (task.getClass().equals(Task.class)) {
+            type = TASK;
+        } else if (task.getClass().equals(Epic.class)) {
+            type = EPIC;
+        } else if (task.getClass().equals(Subtask.class)) {
+            type = SUBTASK;
+            epicId = String.valueOf(((Subtask) task).getEpicId());
+        }
```
Можно реализовать обыкновенные геттеры в классах задач, которые будут возвращать нужный тип в зависимости от класса где он реализован

Ivanov-s-tmn commented on src/main/service/HistoryManager.java:
```diff
@@ -10,7 +10,7 @@ public interface HistoryManager {
 
     //Вернуть историю просмотренных задач
-    List<Task> getHistory();
+    List<Task> getAll();
```
Лучше все-таки вернуть название `getHistory()`, метод возвращает историю

Ivanov-s-tmn commented on src/main/service/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,217 @@
+            while (br.ready()) {
+                String line = br.readLine();
+                Task task = fromString(line);
+                assert task != null : "Задача пустая";
```
Такие ассерты в коде не используются, они по сути не защищают ни от чего. Простая проверка гораздо безопаснее, ее можно реализовать в методе `fromString()`, а вообще, лучше обработать случай чтения пустой строки

Ivanov-s-tmn commented on src/main/service/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,217 @@
+                    subtaskStorage.put(subtask.getId(), subtask);
+                    Epic epic = epicStorage.get(subtask.getEpicId());
+                    epic.setSubtasks(subtask.getId());
+                    epicStorage.put(subtask.getEpicId(), epic);
```
Нет необходимости заменять эпик, `epicStorage.get(subtask.getEpicId());` возвращает ссылку на исходный объект в памяти

Ivanov-s-tmn commented on test/service/FileBackedTaskManagerTest.java:
Сохранение задач в файловый менеджер работает при вызове методов сохранения/удаления/обновления, так как эта логика вшита в его переопределенные методы таск менеджера. Соответственно в тестах для создания, удаления и обновления задач нужно пользоваться только методами файлового менеджера, а выгружать данные из файла с помощью loadFromFile(), как пример тест-кейса, который нужно реализовать:
1. Создать файловый менеджер 
2. Заполнить задачами 
3. Создать новый файловый менеджер из этого же файла через loadFromFile()
4. Проверить, что новый менеджер корректно восстановил состояние из файла (списки задач до и после выгрузки должны быть идентичны)

devices11 replied:
Опять я с тестами увлекся ) Забываюсь, что нужно чисто проверки класса делать. Убрал ненужные тесты и оставил всего три, сохранение пустого файла, сохранение задач и вычитка задач.

Commit:
1. Добавлены гетеры типа задачи в модели задач
2. Вынесены в утилитарный класс методы преобразования задачи в строку и обратно
3. Рефакторинг по замечания
4. Исправлены тесты для FileBackedTaskManager

Ivanov-s-tmn commented on test/service/FileBackedTaskManagerTest.java:
```diff
@@ -97,95 +90,29 @@ void createTaskAndSave() {
-                "Файл некорректный");
-
+    void readFile() {
+        List<Task> tasks = new ArrayList<>(taskManagerFromFile.getAllTask());
+        List<Epic> epics = new ArrayList<>(taskManagerFromFile.getAllEpic());
+        List<Subtask> subtasks = new ArrayList<>(taskManagerFromFile.getAllSubtask());
+
+        assertEquals(3, tasks.size() + epics.size() + subtasks.size(), "Количество задач некорректно");
+
+        assertEquals(tasks.getFirst().getId(), 1, "id задачи некорректен");
+        assertEquals(tasks.getFirst().getTitle(), "Название таски 1", "Название задачи некорректно");
+        assertEquals(tasks.getFirst().getStatus(), NEW, "Статус задачи некорректен");
+        assertEquals(tasks.getFirst().getDescription(), "Описание таски 1", "Описание задачи некорректно");
+
+        assertEquals(epics.getFirst().getId(), 2, "id эпика некорректен");
+        assertEquals(epics.getFirst().getTitle(), "Название эпика 1", "Название эпика некорректно");
+        assertEquals(epics.getFirst().getStatus(), NEW, "Статус эпика некорректен");
+        assertEquals(epics.getFirst().getDescription(), "Описание эпика 1", "Описание эпика некорректно");
+
+        assertEquals(subtasks.getFirst().getId(), 3, "id подзадачи некорректен");
+        assertEquals(subtasks.getFirst().getTitle(), "Название сабтаски 1", "Название подзадачи некорректно");
+        assertEquals(subtasks.getFirst().getStatus(), NEW, "Статус задачи некорректен");
+        assertEquals(subtasks.getFirst().getDescription(), "Описание сабтаски 1", "Описание подзадачи некорректно");
+        assertEquals(subtasks.getFirst().getEpicId(), 2, "id эпика некорректно");
```
Смотри, не совсем тот тест, который я хотел видеть. У нас добавился функционал сохранения в файл и соответственно восстановления состояния из файла. Менеджер сохраняет задачи в файл при вызове методов создания/обновления/удаления. А восстановление состояния происходит через метод loadFromFile() - он создает новый менеджер, читает файл и заполняет данными из файла. Именно это и нужно проверить в тесте. 
```java
//Для начала создать файловый менеджер 
FileBackedTaskManager fileManager = new FileBackedTaskManager(file);
заполнить его задачами (создать к примеру по 1 задаче каждого типа)
fileManager.createTask(task);
....
//После этого создать файловый менеджер из этого же файла через loadFromFile()
FileBackedTaskManager backedManager= FileBackedTaskManager.loadFromFile(file);
//и проверить, что списки до и после выгрузки идентичны, как пример
assertEquals(fileManager.getAllTask(), backedManager.getAllTask());
```

devices11 replied:
Добавил тест

Commit:
1. Добавлен тест в FileBackedTaskManagerTest
